transmission_interface/src/transmission_parser.cpp

```diff
@@ -12,68 +12,250 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "transmission_interface/transmission_parser.hpp"
-#include "transmission_interface/transmission_info.hpp"
+#include <transmission_interface/transmission_parser.hpp>
 
-#include <tinyxml2.h>
-#include <stdexcept>
+// ROS
+#include <rclcpp/rclcpp.hpp>
+
+#include <sstream>
 #include <string>
 #include <vector>
 
-namespace
-{
-constexpr const auto kTransmissionTag = "transmission";
-constexpr const auto kJointTag = "joint";
-constexpr const auto kHardwareInterfaceTag = "hardwareInterface";
-}  // namespace
 
 namespace transmission_interface
 {
+static const rclcpp::Logger g_logger(rclcpp::get_logger("transmission_parser"));
 
 std::vector<TransmissionInfo> parse_transmissions_from_urdf(const std::string & urdf)
 {
-  if (urdf.empty()) {
-    throw std::runtime_error("empty URDF passed in to transmission parser");
+  TransmissionParser parser;
+  std::vector<TransmissionInfo> transmissions;
+  if (!parser.parse(urdf, transmissions)) {
+    throw std::runtime_error("failed to parse transmissions in URDF");
   }
+  return transmissions;
+}
+
 
-  tinyxml2::XMLDocument doc;
+bool TransmissionParser::parse(
+  const std::string & urdf,
+  std::vector<TransmissionInfo> & transmissions)
+{
+  // initialize TiXmlDocument doc with a string
+  TiXmlDocument doc;
   if (!doc.Parse(urdf.c_str()) && doc.Error()) {
-    throw std::runtime_error("invalid URDF passed in to transmission parser");
+    RCLCPP_ERROR(g_logger, "Can't parse transmissions. Invalid robot description.");
```

Comment:
why not throw an error?

Reply:
This code was rewritten by Bence and I. Only returns exceptions now.